presto-hive/src/main/java/com/facebook/presto/hive/metastore/RecordingHiveMetastore.java

```diff
@@ -0,0 +1,582 @@
+        tablePrivilegesCache.putAll(toMap(recording.getTablePrivileges()));
+    }
+
+    private static <K, V> Cache<K, V> createCache(HiveClientConfig hiveClientConfig)
+    {
+        if (hiveClientConfig.isReplay()) {
+            return CacheBuilder.<K, V>newBuilder()
+                    .build();
+        }
+
+        return CacheBuilder.<K, V>newBuilder()
+                .expireAfterWrite(hiveClientConfig.getRecordingDuration().toMillis(), MILLISECONDS)
+                .build();
+    }
+
+    @Managed
+    public void writeRecording()
+            throws IOException
+    {
+        if (replay) {
+            throw new PrestoException(CONFIGURATION_INVALID, "Cannot write recording in replay mode");
+        }
+
+        Recording recording = new Recording(
+                allDatabases,
+                toPairsList(databaseCache),
+                toPairsList(tableCache),
+                toPairsList(tableStatisticsCache),
+                toPairsList(partitionStatisticsCache),
+                toPairsList(allTablesCache),
+                toPairsList(allViewsCache),
+                toPairsList(partitionCache),
+                toPairsList(partitionNamesCache),
+                toPairsList(partitionNamesByPartsCache),
+                toPairsList(partitionsByNamesCache),
+                toPairsList(rolesCache),
+                toPairsList(databasePrivilegesCache),
+                toPairsList(tablePrivilegesCache));
+        new ObjectMapperProvider().get()
+                .writerWithDefaultPrettyPrinter()
+                .writeValue(new File(recordingPath), recording);
+    }
+
+    private static <K, V> Map<K, V> toMap(List<Pair<K, V>> pairs)
+    {
+        return pairs.stream()
+                .collect(ImmutableMap.toImmutableMap(Pair::getKey, Pair::getValue));
+    }
+
+    private static <K, V> List<Pair<K, V>> toPairsList(Cache<K, V> cache)
+    {
+        return cache.asMap().entrySet().stream()
+                .map(entry -> new Pair<>(entry.getKey(), entry.getValue()))
+                .collect(toImmutableList());
+    }
+
+    @Override
+    public Optional<Database> getDatabase(String databaseName)
+    {
+        return loadValue(databaseCache, databaseName, () -> delegate.getDatabase(databaseName));
+    }
+
+    @Override
+    public List<String> getAllDatabases()
+    {
+        if (replay) {
+            return allDatabases.orElseGet(delegate::getAllDatabases);
+        }
+
+        List<String> result = delegate.getAllDatabases();
+        allDatabases = Optional.of(result);
+        return result;
+    }
+
+    @Override
+    public Optional<Table> getTable(String databaseName, String tableName)
+    {
+        return loadValue(tableCache, table(databaseName, tableName), () -> delegate.getTable(databaseName, tableName));
+    }
+
+    @Override
+    public Set<ColumnStatisticType> getSupportedColumnStatistics(Type type)
+    {
+        return delegate.getSupportedColumnStatistics(type);
```

Comment:
Add a comment why this is not using `loadValue`

Reply:
Actually it should so that we don't need to plug metastore underneath (e.g: Glue)